core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * // create a registry
+ * ObservationRegistry observationRegistry = ObservationRegistry.create();
+ *
+ * // add the tracing handlers
+ * observationRegistry.observationConfig().observationHandler(
+ *         new FirstMatchingCompositeObservationHandler(tracingHandlers));
+ *
+ * // add the decorator
+ * WebClient.builder()
+ *          .decorator(MicrometerObservationClient.newDecorator(observationRegistry))
+ * ...
+ * }</pre>
+ */
+@UnstableApi
+public final class ObservationClient extends SimpleDecoratingHttpClient {
+
+    /**
+     * Creates a new micrometer observation integrated {@link HttpClient} decorator
+     * using the specified {@link ObservationRegistry}.
+     */
+    public static Function<? super HttpClient, ObservationClient> newDecorator(
+            ObservationRegistry observationRegistry) {
+        requireNonNull(observationRegistry, "observationRegistry");
+        return delegate -> new ObservationClient(delegate, observationRegistry, null);
+    }
+
+    /**
+     * Creates a new micrometer observation integrated {@link HttpClient} decorator
+     * using the specified {@link ObservationRegistry} and {@link ObservationConvention}.
+     */
+    public static Function<? super HttpClient, ObservationClient> newDecorator(
+            ObservationRegistry observationRegistry,
+            @Nullable ObservationConvention<ClientObservationContext> observationConvention) {
+        requireNonNull(observationRegistry, "observationRegistry");
+
+        return delegate -> new ObservationClient(delegate, observationRegistry,
+                                                 observationConvention);
+    }
+
+    private final ObservationRegistry observationRegistry;
+
+    @Nullable
+    private final ObservationConvention<ClientObservationContext> httpClientObservationConvention;
+
+    private ObservationClient(
+            HttpClient delegate, ObservationRegistry observationRegistry,
+            @Nullable ObservationConvention<ClientObservationContext> httpClientObservationConvention) {
+        super(delegate);
+        this.observationRegistry = requireNonNull(observationRegistry, "observationRegistry");
+        this.httpClientObservationConvention = httpClientObservationConvention;
+    }
+
+    @Override
+    public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+        final RequestHeadersBuilder newHeaders = req.headers().toBuilder();
+        final ClientObservationContext clientObservationContext =
+                new ClientObservationContext(ctx, newHeaders, req);
+        final Observation observation = HttpClientObservationDocumentation.OBSERVATION.observation(
+                httpClientObservationConvention, DefaultHttpClientObservationConvention.INSTANCE,
+                () -> clientObservationContext, observationRegistry).start();
+        final HttpRequest newReq = req.withHeaders(newHeaders);
```

Comment:
Question: Do the headers are added to the `newHeaders` when `start()` is called? (I have tried to find it, but I couldn't. 😓 )

Reply:
Ah ha got it. Thanks for the explanation. 👍